### PR TITLE
RoR group project: Recipe app

### PR DESCRIPTION
**Pull request by sanieni6**
# RoR group project: Recipe app🤓 
This is a group capstone project where we create a RecipeApp that will help you to keep a record of your recipes. Everything will be based on the UML class diagram and the wireframes provides by Microverse . 
## Members of the group: 🧑‍🤝‍🧑 
@Cabraham1 
@sanieni6 

## Summary 🏠 
+ Create the DB following the provided UML design ✅ 
+ Use the ORM model to create  full stack project ✅ 
+ Unit tests and integration test ✅ 
+ Recreate the database on PostgreSQL using the provided UML design ✅ 
## What we learned ☑️ 

- Install Ruby on Rails framework.
- Use controllers to handle requests and render empty views.
- Understand Rails naming conventions.
- Use layouts and templates for shared content.
- Use database migration files to maintain database schema.
- Use validations for models.
- Secure app from n+1 problems.
- Understand what ORM is.
- Write SQL queries with ActiveRecord.
- Set up associations between models.
- Build a web app that requires the user to log in.
- Describe the difference between authorization and authentication.
- Use Devise gem for authentication.
- Understand how sessions and cookies can support authentication.
- Limit access to web app resources based on authorization rules.
- Write unit tests using Rspec
- Write integration tests with capybara gem

 ## 📷 Screenshots
image
image
image
image_720
image

image

image_720

### CODE REVIEW

File: app/controllers/generalshoppinglist_controller.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        }
+      end
+    end
+
+    missing_foods
+  end
+
+  def calculate_total_price(foods)
+    foods.sum { |food| food[:price] * food[:quantity] }
+  end
```
Comment (tugsem) [inline app/controllers/generalshoppinglist_controller.rb:37]: - [ ] As per the [skinny controller rule](https://dev.to/kputra/rails-skinny-controller-skinny-model-5f2k), please keep the necessary logic only in your controllers to make a request. I recommend you to move these functions into your model or helpers and so you can test them without making a request.

**Review by tugsem — CHANGES_REQUESTED** [review]
##  STATUS: REQUIRED CHANGES ♻️
Hello  @sanieni6  and @Cabraham1 👋 , 

Good job so far! 👏 
There are some issues that you still need to work on to go to the next project but you are almost there!

Alt Text
## Highlights 🚀 
✔️professional documentation
✔️correct use of gitflow
✔️app works as expected
✔️ code is clear and easy to understand
✔️ design wireframe almost identical
✔️ tests are written with capybara
## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by sanieni6** [comment]
# 🟢  Changes done: Recipe datils for unauthorized person on public recipes

- Case: When the current user is the owner
image
- Case: When the current user is not the owner
image

**Review by mukhammadrizooff — APPROVED** [review]
Hi @TEAM, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉

 

## Highlights
- No linter errors ✔️ 
- Correct GitFlow ✔️ 
- Good implementations ✔️
- Descriptive `README` file ✔️

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**